x-pack/plugins/uptime/server/lib/adapters/monitors/elasticsearch_monitors_adapter.ts

```diff
@@ -97,32 +86,41 @@ export class ElasticsearchMonitorsAdapter implements UMMonitorsAdapter {
     };
 
     const result = await this.database.search(request, params);
-    const buckets = get(result, 'aggregations.timeseries.buckets', []);
+    const buckets = dropLatestBucket(get(result, 'aggregations.timeseries.buckets', []));
 
-    return buckets.map(
-      ({
-        key,
-        max_content,
-        duration: { avg, max, min },
-        max_write_request,
-        max_validate,
-        max_tcp_rtt,
-        max_response,
-        max_total,
-        status,
-        doc_count,
-      }: any) => ({
-        maxContent: formatChartValue(key, max_content),
-        maxWriteRequest: formatChartValue(key, max_write_request),
-        maxValidate: formatChartValue(key, max_validate),
-        maxTcpRtt: formatChartValue(key, max_tcp_rtt),
-        maxResponse: formatChartValue(key, max_response),
-        maxTotal: formatChartValue(key, max_total),
-        avgDuration: formatChartValue(key, { value: avg }),
-        maxDuration: formatChartValue(key, { value: max }),
-        minDuration: formatChartValue(key, { value: min }),
-        status: formatStatusBuckets(key, status.buckets, doc_count),
-      })
+    let statusMaxCount = 0;
+    let durationMaxValue = 0;
+    const mappedBuckets = buckets.map(bucket => {
+      const x = get(bucket, 'key');
+      const docCount = get(bucket, 'doc_count', 0);
+      statusMaxCount = Math.max(docCount, statusMaxCount);
+      durationMaxValue = Math.max(durationMaxValue, get(bucket, 'duration.max', 0));
+      return {
+        x,
+        durationMin: get(bucket, 'duration.min', null),
+        durationMax: get(bucket, 'duration.max', null),
+        durationAvg: get(bucket, 'duration.avg', null),
+        status: formatStatusBuckets(x, get(bucket, 'status.buckets', []), docCount),
+      };
+    });
+    const counter: MonitorChart = {
+      durationArea: [],
+      durationLine: [],
+      status: [],
+      durationMaxValue,
+      statusMaxCount,
+    };
+    return mappedBuckets.reduce(
```

Comment:
Is map/reduce getting us much here?

As a fan of FP (and former lisper) I like MR, but it seems to be interfering with code clarity here. We have an entire intermediate set of variables created in the `map` phase that only exists to be discarded and renamed in the reduce phase.

This would be more clear to read using a simple `forEach` that mutated the `counter` object. Something like the untested code below:

```javascript
const result: MonitorChart = {
    durationArea: [],
    durationLine: [],
    status: [],
    durationMaxValue: 0,
    statusMaxCount: 0,
};

buckets.forEach(bucket => {
  const x = get(bucket, 'key');
  const docCount = get(bucket, 'doc_count', 0);
  result.statusMaxCount = Math.max(docCount, result.statusMaxCount);
  result.durationMaxValue = Math.max(result.durationMaxValue, get(bucket, 'duration.max', 0));
  
  result.durationArea.push({ x, y0: get(bucket, 'duration.min', null), y:  get(bucket, 'duration.max', null)});
  result.durationLine.push({ x, y: get(bucket, 'duration.avg', null) });
  result.status.push(formatStatusBuckets(x, get(bucket, 'status.buckets', []), docCount));
});
return result;
```


Alternatively, if you did want to use map reduce, I'd skip the `map` phase completely. The goal here is to turn a list (of buckets) into a single value. That corresponds to a reduce. The code as it stands uses a map phase to expose an intermediate state. However, that intermediate state is, I would argue, confusing, because it doesn't do much other than rename variables. You could do something similar to the `forEach` above, but use `reduce` instead. Like (also untested):

```javascript
const result: MonitorChart = {
    durationArea: [],
    durationLine: [],
    status: [],
    durationMaxValue: 0,
    statusMaxCount: 0,
};

return buckets.reduce( (bucket, acc) => {
  const x = get(bucket, 'key');
  const docCount = get(bucket, 'doc_count', 0);
  result.statusMaxCount = Math.max(docCount, result.statusMaxCount);
  result.durationMaxValue = Math.max(result.durationMaxValue, get(bucket, 'duration.max', 0));
  
  result.durationArea.push({ x, y0: get(bucket, 'duration.min', null), y:  get(bucket, 'duration.max', null)});
  result.durationLine.push({ x, y: get(bucket, 'duration.avg', null) });
  result.status.push(formatStatusBuckets(x, get(bucket, 'status.buckets', []), docCount));
  return result;
}, result);
```

Reply:
First of all - thank you for taking the time to write all this out.

Secondly I agree with your points. Sometimes we are focused on the code's output so much we forget that others will need to read it in the future, and we implement the very first solution that comes to mind. I think your first solution is best. If we need to declare an accumulator variable before the functions no matter what, it makes sense to access it from the simplest possible block, which in this case would be `forEach`.

Please review https://github.com/elastic/kibana/pull/30561/commits/3ccdd1dc85e8107a6d01a5808d84e7d2aac6635c when you have a chance.